src/lib.rs

```diff
@@ -155,10 +166,37 @@ pub fn init_with_rng<T: Rng>(rng: &mut T) -> Result<(), Error> {
 }
 
 /// Produces a 256-bit crypto hash out of the provided `data`.
-pub fn hash(data: &[u8]) -> [u8; 32] {
+pub fn hash(data: &[u8]) -> [u8; HASH_BYTES] {
     hashing_impl::sha3_256(data)
 }
 
+/// Uses Scrypt key derivation to securely derive a byte vector from a `password` and `salt`.
+/// `output` may be used to construct any of the keys in this library.
+///
+/// `output` must satisfy the following condition: `output.len() > 0 && output.len() <= (2^32 - 1) *
+/// 32`.
+///
+/// # Example
+///
+/// ```
+/// use safe_crypto::*;
+///
+/// let password = b"password!";
+/// let salt = b"salt!";
+/// let mut output = [0; SYMMETRIC_KEY_BYTES];
+///
+/// derive_bytes(password, salt, &mut output).unwrap();
+/// let key1 = SymmetricKey::from_bytes(output);
+/// derive_bytes(password, salt, &mut output).unwrap();
+/// let key2 = SymmetricKey::from_bytes(output);
+///
+/// assert_eq!(key1, key2);
+/// ```
+pub fn derive_bytes(password: &[u8], salt: &[u8], output: &mut [u8]) -> Result<(), Error> {
+    let params = ScryptParams::new(15, 8, 1).map_err(|_| Error::DeriveKey)?;
```

Comment:
How did we get these parameters ? Hope they are not maximum complexity (else will be very sluggish on mobiles/weaker processors/smaller RAMs. We normall use the Optimal thingy (the recommended stuff from rust/lib sodium, so pls confirm it's in the same ballpark as that.

Would also be nice to make these constants and explain why we are choosing them in the comments above them (these are recommended optimals OR we did some testing and think these are fine because they take x cpu cycles and y amt of memory etc. <- along those lines)

Reply:
Yeah, those are the recommended values by the scrypt crate.